cmd/kubeadm/app/util/staticpod/utils.go

```diff
@@ -372,7 +375,7 @@ func ManifestFilesAreEqual(path1, path2 string) (bool, error) {
 	if bytes.Equal(hash1, hash2) {
 		return true, nil
 	}
-	klog.V(4).Infof("Pod manifest files diff:\n%s\n", cmp.Diff(pod1, pod2))
+	klog.V(4).Infof("Pod manifest files diff:\n%s\n", cmp.Diff(pod2, pod1))
```

Comment:
let's move this to the place the function is called:
https://github.com/pacoxu/kubernetes/blob/c0ad5e545e528c55c4297a220ddfbe38d9292aef/cmd/kubeadm/app/phases/upgrade/staticpods.go#L220

new function signature:
```
func ManifestFilesAreEqual(path1, path2 string) (bool, string, error) {
```

2nd argument is the diff, the caller can print its contents:
```
if bytes.Equal(hash1, hash2) {
	return true, "", nil
}
return false, cmp.Diff(pod2, pod1), nil
```

we probably have this mistake already in some places, but the idea is not to call klog from utils.
klog can be called from places where we already print things like `app/phases/upgrade/staticpods.go` or even better `app/phases/` is backend and should not print anything. all printing must only be done from CLI code...but that's a big refactor.

Reply:
It makes sense. Updated.

Reply:
@neolit123 @pacoxu I lost it...
But why do we need to change `cmp.Diff(pod1, pod2)` to `cmp.Diff(pod2, pod1)`?
If I understand correctly, pod1 is generated from currentManifestPath, and pod2 is generated from newManifestPath.

According the description of `cmp.Diff`:
```go
a "-" prefix indicates an element removed from x,
a "+" prefix to indicates an element added from y
func Diff(x, y interface{}, opts ...Option) string {
```

Shouldn't we show what the upgraded new manifest has -/+ contents compared to the current old manifest?

Reply:
@pacoxu 's output showed new defaulted fields with `+ foo` ??
i.e. the 1.29 manifest after upgrade had these strange defaulted fields.
and pod1, pod2 was not correct...

i think we should just get rid of cmp.Diff()'s usage for this code and try difflib 
https://github.com/kubernetes/kubernetes/pull/120602#discussion_r1323164529

if nothing works just drop the whole diff code and unit tests that @pacoxu added.
we are currently flaking k/k master...